packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -0,0 +1,252 @@
+import type {
+  Metadata,
+  ResolvedMetadata,
+  ResolvingMetadata,
+} from './types/metadata-interface'
+import type { AbsoluteTemplateString } from './types/metadata-types'
+import type { Viewport } from './types/extra-types'
+import { createDefaultMetadata } from './default-metadata'
+import { resolveOpenGraph } from './resolve-opengraph'
+import { resolveTitle } from './resolve-title'
+
+const viewPortKeys = {
+  width: 'width',
+  height: 'height',
+  initialScale: 'initial-scale',
+  minimumScale: 'minimum-scale',
+  maximumScale: 'maximum-scale',
+  viewportFit: 'viewport-fit',
+} as const
+
+type Item =
+  | {
+      type: 'layout' | 'page'
+      // A number that represents which layer or routes that the item is in. Starting from 0.
+      // Layout and page in the same level will share the same `layer`.
+      layer: number
+      mod: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path: string
+    }
+  | {
+      type: 'icon'
+      // A number that represents which layer the item is in. Starting from 0.
+      layer: number
+      mod?: () => Promise<{
+        metadata?: Metadata
+        generateMetadata?: (
+          props: any,
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path?: string
+    }
+
+// Merge the source metadata into the resolved target metadata.
+function merge(
+  target: ResolvedMetadata,
+  source: Metadata,
+  templateStrings: {
+    title: string | null
+    openGraph: string | null
+    twitter: string | null
+  }
+) {
+  let updatedStashedTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: templateStrings.title,
+  }
+  let updatedStashedOpenGraphTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: templateStrings.openGraph,
+  }
+  let updatedStashedTwitterTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: templateStrings.twitter,
+  }
+
+  for (const key_ in source) {
+    const key = key_ as keyof Metadata
+
+    switch (key) {
+      case 'other': {
+        target.other = { ...target.other, ...source.other }
+        break
+      }
+      case 'title': {
+        updatedStashedTitle = resolveTitle(templateStrings.title, source.title)
+        target.title = updatedStashedTitle
+        break
+      }
+      case 'openGraph': {
+        if (typeof source.openGraph !== 'undefined') {
+          target.openGraph = {
+            ...resolveOpenGraph(source.openGraph),
+          }
+          if (source.openGraph && 'title' in source.openGraph) {
+            updatedStashedOpenGraphTitle = resolveTitle(
+              templateStrings.openGraph,
+              source.openGraph.title
+            )
+            target.openGraph.title = updatedStashedOpenGraphTitle
+          }
```

Comment:
I would pass the og title template into `resolveOpenGraph` and do the title merging inside it. If you refactor `resolveTitle` to `mergeTitle` just resolve openGraph first without excluding title, but normalize to AbsoluteTemplateString, then merge the source.openGraph.title using the passed in template